libs/ledger-live-common/src/hw/actions/installLanguage.ts

```diff
@@ -0,0 +1,340 @@
+import { ConnectManagerTimeout } from "../../errors";
+import { currentMode } from "./app";
+import {
+  DisconnectedDevice,
+  DisconnectedDeviceDuringOperation,
+} from "@ledgerhq/errors";
+import { getDeviceModel } from "@ledgerhq/devices";
+import { Language } from "../../types/languages";
+
+type State = {
+  isLoading: boolean;
+  requestQuitApp: boolean;
+  unresponsive: boolean;
+  languageInstallationRequested?: boolean;
+  installingLanguage?: boolean;
+  languageInstalled?: boolean;
+  device: Device | null | undefined;
+  deviceInfo: DeviceInfo | null | undefined;
+  error: Error | null | undefined;
+  progress?: number;
+};
+
+type InstallLanguageAction = Action<Language, State, boolean | undefined>;
+
+const mapResult = ({ languageInstalled }: State) => languageInstalled;
+
+type Event =
+  | InstallLanguageEvent
+  | {
+      type: "error";
+      error: Error;
+    }
+  | {
+      type: "deviceChange";
+      device: Device | null | undefined;
+    };
+
+const getInitialState = (device?: Device | null | undefined): State => ({
+  isLoading: !!device,
+  requestQuitApp: false,
+  unresponsive: false,
+  device,
+  deviceInfo: null,
+  error: null,
+});
+
+const reducer = (state: State, e: Event): State => {
+  switch (e.type) {
+    case "unresponsiveDevice":
+      return { ...state, unresponsive: true, isLoading: false };
+
+    case "deviceChange":
+      return getInitialState(e.device);
+
+    case "error":
+      return {
+        ...getInitialState(state.device),
+        error: e.error,
+        isLoading: false,
+      };
+    case "appDetected":
+      return {
+        ...state,
+        unresponsive: false,
+        requestQuitApp: true,
+        isLoading: false,
+      };
+    case "devicePermissionRequested":
+      return {
+        ...state,
+        unresponsive: false,
+        languageInstallationRequested: true,
+        isLoading: false,
+      };
+    case "languageInstalled":
+      return {
+        ...state,
+        unresponsive: false,
+        languageInstallationRequested: false,
+        isLoading: false,
+        installingLanguage: false,
+        languageInstalled: true,
+      };
+
+    case "progress":
+      return {
+        ...state,
+        unresponsive: false,
+        languageInstallationRequested: false,
+        isLoading: false,
+        installingLanguage: true,
+        progress: e.progress,
+      };
+  }
+};
+
+const implementations = {
+  // in this paradigm, we know that deviceSubject is reflecting the device events
+  // so we just trust deviceSubject to reflect the device context (switch between apps, dashboard,...)
+  event: ({ deviceSubject, installLanguage, language }) =>
+    deviceSubject.pipe(
+      debounceTime(1000),
+      switchMap((d) => installLanguage(d, language))
+    ),
+  // in this paradigm, we can't observe directly the device, so we have to poll it
+  polling: ({ deviceSubject, installLanguage, language }) =>
+    Observable.create((o) => {
```

Comment:
let's use `new Observable((o) => {...})` as `create` is going to be deprecated

Reply:
good catch. Done